### PR TITLE
perf: Optimize LOH string allocations in Xaml generator

### LINKED ISSUE #11515: Avoid LOH string allocations in XamlCodeGenerator
### Current behavior

image

### Expected behavior

_No response_

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

None

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
Found LOH string allocations in a trace:

image

So now, we are not allocating these large strings and letting Roslyn do the right thing with `CopyTo` and `ToString(TextSpan)` overrides.

closes #11515

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 1]
```diff
@@ -310,7 +311,7 @@ public string GenerateFile()
 #if DEBUG
 					Console.WriteLine("Skipping unmodified file {0}".InvariantCultureFormat(_fileDefinition.FilePath));
 #endif
-					return File.ReadAllText(outputFile);
+					return SourceText.From(File.ReadAllText(outputFile), Encoding.UTF8);
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:314]: I didn't care about this code path for now. We will probably remove if/when we are using Roslyn-based generation only.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs [hunk 2]
```diff
@@ -959,7 +960,8 @@ void RegisterForXamlFile(XamlFileDefinition file, string url)
 				}
 			}
 
-			return writer.ToString();
+			Debugger.Launch();
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:963]:
```suggestion
```

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs (hunk 1 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:314]: Agreed.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/Utils/StringBuilderBasedSourceText.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+	public override Encoding Encoding => Encoding.UTF8;
+
+	public override int Length => _builder.Length;
+
+	public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count)
+		=> _builder.CopyTo(sourceIndex, destination, destinationIndex, count);
+
+	public override string ToString(TextSpan span)
+		=> _builder.ToString(span.Start, span.Length);
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/Utils/StringBuilderBasedSourceText.cs:33]: Should we be caching this particular output? Is it used often?

**Review by jeromelaban — APPROVED** [review]

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/Utils/StringBuilderBasedSourceText.cs (hunk 3 above)
Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/Utils/StringBuilderBasedSourceText.cs:33]: I tried adding Debugger. Launch here if the span corresponds to the whole string and I didn't see it being hit. 

For the most part, Roslyn calls into CopyTo, so I think best to keep it simple and optimize as we see issues.

**Pull request merged**